oauth/src/OAuth/OAuth2.hs

```diff
@@ -0,0 +1,220 @@
+  deriving stock (Eq, Generic, Show)
+
+newtype RefreshToken = RefreshToken Text
+  deriving newtype (JSON.ToJSON, JSON.FromJSON)
+  deriving stock (Eq, Show)
+
+data AuthenticationRequest = AuthenticationRequest
+  { clientId     :: Text
+  , clientSecret :: Text
+  , code         :: AuthCode
+  , grantType    :: Text
+  , nonce        :: Text
+  , redirectUri  :: Text
+  , state        :: [(Text, Text)]
+  }
+  deriving stock (Eq, Generic, Show)
+
+instance JSON.ToJSON AuthenticationRequest where
+  toJSON     = JSON.genericToJSON oAuth2JsonOptions
+  toEncoding = JSON.genericToEncoding oAuth2JsonOptions
+
+instance JSON.FromJSON AuthenticationRequest where
+  parseJSON = JSON.genericParseJSON oAuth2JsonOptions
+
+data AuthenticationResponse = AuthenticationResponse
+  { accessToken  :: Text
+  , expiresIn    :: Integer
+  , idToken      :: Text
+  , refreshToken :: Text
+  , scope        :: Text
+  , tokenType    :: Text
+  }
+  deriving stock (Eq, Generic, Show)
+
+instance JSON.FromJSON AuthenticationResponse where
+  parseJSON = JSON.genericParseJSON oAuth2JsonOptions
+
+data OpenIdPayload = OpenIdPayload
+  { email         :: Text
+  , emailVerified :: Bool
+  }
+  deriving stock (Eq, Generic, Show)
+
+instance JSON.FromJSON OpenIdPayload where
+  parseJSON = JSON.genericParseJSON oAuth2JsonOptions
+
+data RefreshAccessTokenRequest = RefreshAccessTokenRequest
+  { clientId     :: Text
+  , clientSecret :: Text
+  , grantType    :: Text
+  , refreshToken :: RefreshToken
+  }
+  deriving stock (Eq, Generic, Show)
+
+instance JSON.ToJSON RefreshAccessTokenRequest where
+  toJSON     = JSON.genericToJSON oAuth2JsonOptions
+  toEncoding = JSON.genericToEncoding oAuth2JsonOptions
+
+instance JSON.FromJSON RefreshAccessTokenRequest where
+  parseJSON = JSON.genericParseJSON oAuth2JsonOptions
+
+data RefreshAccessTokenResponse = RefreshAccessTokenResponse
+  { accessToken :: Text
+  , expiresIn   :: Integer
+  , scope       :: Text
+  , tokenType   :: Text
+  }
+  deriving stock (Eq, Generic, Show)
+
+instance JSON.FromJSON RefreshAccessTokenResponse where
+  parseJSON = JSON.genericParseJSON oAuth2JsonOptions
+
+authorizationRequestUrl :: Credentials -> [Text] -> Text
+authorizationRequestUrl Credentials{..} scopes =
+  Text.decodeUtf8 $ "https://accounts.google.com/o/oauth2/v2/auth" <> query
+  where
+    query :: ByteString
+    query = URI.renderSimpleQuery True $ second Text.encodeUtf8 <$>
+      [ ("access_type",   "offline")
+      , ("client_id",     clientId)
+      , ("redirect_uri",  fromJust $ safeHead redirectUris)
+      , ("response_type", "code")
+      , ("scope",         Text.intercalate " " $ ["openid", "email"] <> scopes)
+      ]
+
+authenticate
+  :: forall env . Env env
+  => AuthCode
+  -> RIO env AuthenticationResponse
+authenticate code = do
+  Credentials{..} <- asks (getField @"googleOAuth2Credentials")
+  nonce           <- liftIO $ Nonce.withGenerator Nonce.nonce128urlT
```

Comment:
@snusnu  Previously the `nonce` value was a function argument. Let me know if there is any special reason for the previous setup. Otherwise, I prefer this scenario where it's internally provided within the function body.

Reply:
@epicallan I can't remember a specific reason. If you prefer that way, that's fine with me.